### PR TITLE
[EXPERIMENTAL] Part 3: Faster Qblox integration

### PR DESCRIPTION
**Pull request by hamidelmaazouz**
The [original PR](https://github.com/oqc-community/qat/pull/157) was deemed too complex and we decided to split it in three main parts. The [first part](https://github.com/oqc-community/qat/pull/216) is the pass infrastructure, the [second part](https://github.com/oqc-community/qat/pull/217) is control-flow infrastructure, and the third part leverages previous work for a faster Qblox integration. This PR covers only the third part.

- Introduces a new legalisation pass for Qblox. This pass builds upon the target-agnostic `TILegalisationPass` to finalise iterator bound conversion prior to codegen.
- The PreCodegenPass allows one to properly initialise condegen contexts prior to invoking the emitter. This has the role of early register and label allocation for the loop nests in the builder as well as for the acquisition instructions.
- New Qblox codegen becomes a mere DFS traversal on the CFG to allow for more flexibility and control.
- Code generation is a much more complex topic, but CFG traversal technique introduced here is enough for the current iteration.

### CODE REVIEW

File: src/qat/purr/backends/qblox/live.py [hunk 1]
```diff
@@ -103,3 +131,182 @@ def _execute_on_hardware(
...
+        pipeline = (
+            PassManager()
+            | SweepDecomposition()
+            | RepeatSanitisation()
+            | ScopeSanitisation()
+            | ReturnSanitisation()
+            | ScopeSanitisationValidation()
+            | ReturnSanitisationValidation()
+            | RepeatSanitisationValidation()
+            | TriagePass()
```
Comment (daria-oqc) [inline src/qat/purr/backends/qblox/live.py:163]: Love this, great work!

File: src/qat/purr/backends/qblox/live.py [hunk 2]
```diff
@@ -103,3 +131,182 @@ def _execute_on_hardware(
...
+            res_mgr = ResultManager()
+            self.run_pass_pipeline(builder, res_mgr, self.model)
+            packages = NewQbloxEmitter().emit_packages(builder, res_mgr, self.model)
+            self.model.control_hardware.set_data(packages)
+            playback_results: Dict[str, np.ndarray] = (
+                self.model.control_hardware.start_playback(None, None)
+            )
+
+            # Post execution step needs a lot of work
+            # TODO - Robust batching analysis (as a pass !)
```
Comment (daria-oqc) [inline src/qat/purr/backends/qblox/live.py:180]: 👍🏻

File: src/qat/purr/backends/qblox/live.py [hunk 3]
```diff
@@ -103,3 +131,182 @@ def _execute_on_hardware(
...
+            # TODO - Lowerability analysis pass
+            # TODO - A generic loop nest model. Sth similar to SweepIterator but does not mixin injection stuff
+
+            triage_result: TriageResult = res_mgr.lookup_by_type(TriageResult)
+            acquire_map = triage_result.acquire_map
+            pp_map = triage_result.pp_map
+            sweeps = triage_result.sweeps
+
+            def create_sweep_iterator():
```
Comment (daria-oqc) [inline src/qat/purr/backends/qblox/live.py:189]: Do we need a separate function for this? It's only called once as far as I know.

**Review by daria-oqc — APPROVED** [review]

File: src/qat/purr/backends/qblox/live.py (hunk 3 above)
Comment (hamidelmaazouz) [reply src/qat/purr/backends/qblox/live.py:189]: post execution steps are still fuzzy for qblox, will improve on that indeed

File: src/qat/purr/backends/qblox/live.py [hunk 4]
```diff
@@ -1,24 +1,45 @@
...
+    ScopeSanitisationValidation,
+)
+from qat.ir.pass_base import InvokerMixin, PassManager
+from qat.ir.result_base import ResultManager
 from qat.purr.backends.live import LiveDeviceEngine, LiveHardwareModel
 from qat.purr.backends.live_devices import ControlHardware
-from qat.purr.backends.qblox.codegen import QbloxEmitter
+from qat.purr.backends.qblox.codegen import NewQbloxEmitter, QbloxEmitter
 from qat.purr.backends.utilities import get_axis_map
+from qat.purr.compiler.config import InlineResultsProcessing
```
Comment (keriksson-rosenqvist) [inline src/qat/purr/backends/qblox/live.py:23]: Legacy path. I believe this should be imported from `compiler_config.config`

**Review by keriksson-rosenqvist — APPROVED** [review]
LGTM.
Just a minor on legacy imports

**Comment by keriksson-rosenqvist** [comment]
Why is this not merging to main?

File: src/qat/model/component.py
Comment (keriksson-rosenqvist) [inline src/qat/model/component.py:1]: This isn't actually part of this PR is it? :confused:

**Review by keriksson-rosenqvist — COMMENTED** [review]
Looks like there's some weirdness going on with commits here.

File: src/qat/purr/backends/qblox/live.py (hunk 3 above)
Comment (hamidelmaazouz) [reply src/qat/purr/backends/qblox/live.py:192]: Happy with this for now as I have greater plans for this part and I will need to recalibrate my reasoning based on internal use cases.

**Comment by hamidelmaazouz** [comment]
> Why is this not merging to main?

the three PRs were stacked to help with review, now that the first two are merged, I discovered a few issues with semantic analyses. Now that they are fixed, this PR targets main and the diff should be cleaner

File: src/qat/model/component.py
Comment (hamidelmaazouz) [reply src/qat/model/component.py:1]: the three PRs were stacked to help with review, this PR was targetting an old base. it now targets main and the diff should be cleaner

**Pull request merged**